### PR TITLE
Package metadata abstract

### PR DESCRIPTION
**Pull request by shaded-enmity**
Initial package metadata proposal

### CODE REVIEW

File: packages/README.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+```
+         +-------------------+
+IN       |                   |  OUT
+         |  npm metadata     |
++------->+  manifest tool    +------------>
+         |                   |
+         +-------------------+
+```
+
+`IN`put is a `npm` package and `OUT`put is a document in common metadata format. Whenever the `npm` upstream changes its package specification
```
Comment (TomasTomecek) [inline packages/README.md:29]: will we store metadata per version? (dependencies tend to change between major releases)

File: packages/README.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+The language of the common metadata format shall be expressive enough so that we can always express a feature of one package format in terms
+of the common metadata format. Expressivity can be thought of as the degree of *freedom* or *possibilities* that is needed to describe the particular
+metadata component. Let's illustrate this on versioned dependencies - while some package ecosystem can do fine with rather simplistic rules for dependency versioning:
+
+  * `>= 1.0.0`
+  * `1.0.0`
+  * `>= 1.2`
+  
+There also exist much *more complex* dependency variants:
```
Comment (TomasTomecek) [inline packages/README.md:46]: How far do we want to go here? https://docs.npmjs.com/files/package.json#dependencies

File: packages/README.md (hunk 1 above)
Comment (shaded-enmity) [inline packages/README.md:29]: Yes, optimally, the input archive is digested using `sha` or whatever and that is used as an unique identifier within our infrastructure.

File: packages/README.md (hunk 2 above)
Comment (shaded-enmity) [inline packages/README.md:46]: Both this and the Maven dependency specifications are rather horrifying, perhaps we'll need to store the dep-graph in an AST or something like that. I mean, the syntax is complicated but the underlying operating block is _sets_ + _comparison operators_.

Comment (TomasTomecek) [inline packages/README.md:46]: Depends if we want to resolve deps with our code or create a backend task which will ensure that module is installable using packages in our repos.

Comment (shaded-enmity) [inline packages/README.md:46]: Good question, we should dep-solve early in the process; combined with using minimal build roots we can catch many silly mistakes and include them in the metadata - sometimes people assume something is standard when it really isn't:
- people forget to require `six` in Python, assuming it's near standard
- `which`, `find` etc. are used during `pip/gem/npm install`, yet many of these commands are not in base installations (Docker base image for example)
- not to mention `C/C++` compilers, `cmake` etc.

The architecture should be able to deal with all of that by means of static analysis. I have a presentation about this for the next sprint, so we can discuss there more.

**Pull request merged**